From 5ea88361017913cccd3ee460457f61013df2dced Mon Sep 17 00:00:00 2001
From: Etaash Mathamsetty
 <45927311+Etaash-mathamsetty@users.noreply.github.com>
Date: Sun, 4 Jan 2026 01:03:48 -0500
Subject: [PATCH 473/479] HACK: mfplat: Use the MP4 bytestream handler as a
 fallback with mp4 url hint.

---
 dlls/mfplat/main.c | 16 +++++++++++++++-
 1 file changed, 15 insertions(+), 1 deletion(-)

diff --git a/dlls/mfplat/main.c b/dlls/mfplat/main.c
index 1a24d380d0d..db2a39d4099 100644
--- a/dlls/mfplat/main.c
+++ b/dlls/mfplat/main.c
@@ -43,6 +43,7 @@
 #include "strsafe.h"
 #undef INITGUID
 #include "evr.h"
+#include "wine/mfinternal.h"
 /* mfd3d12 guids are not included in mfuuid */
 #define INITGUID
 #undef EXTERN_GUID
@@ -6304,6 +6305,11 @@ static HRESULT resolver_create_gstreamer_handler(IMFByteStreamHandler **handler)
     return CoCreateInstance(&CLSID_GStreamerByteStreamHandler, NULL, CLSCTX_INPROC_SERVER, &IID_IMFByteStreamHandler, (void **)handler);
 }
 
+static HRESULT resolver_create_mpeg4_handler(IMFByteStreamHandler **handler)
+{
+    return CoCreateInstance(&CLSID_MPEG4ByteStreamHandlerPlugin, NULL, CLSCTX_INPROC_SERVER, &IID_IMFByteStreamHandler, (void **)handler);
+}
+
 static HRESULT resolver_get_bytestream_handler(IMFByteStream *stream, const WCHAR *url, DWORD flags,
         IMFByteStreamHandler **handler)
 {
@@ -6337,12 +6343,20 @@ static HRESULT resolver_get_bytestream_handler(IMFByteStream *stream, const WCHA
        this handler for all possible types.
      */
 
+    TRACE("url_ext %s mimeW %s\n", debugstr_w(url_ext), debugstr_w(mimeW));
+
     if (url_ext || mimeW)
     {
         hr = resolver_create_bytestream_handler(stream, flags, mimeW, url_ext, handler);
 
         if (FAILED(hr))
-            hr = resolver_create_gstreamer_handler(handler);
+        {
+            if (SUCCEEDED(resolver_get_bytestream_url_hint(stream, &url_ext)) &&
+                url_ext && (!wcscmp(url_ext, L".mp4") || !wcscmp(url_ext, L".m4v")))
+                hr = resolver_create_mpeg4_handler(handler);
+            else
+                hr = resolver_create_gstreamer_handler(handler);
+        }
     }
 
     CoTaskMemFree(mimeW);
-- 
2.52.0

